sync15-adapter/src/token.rs

```diff
@@ -20,91 +23,106 @@ header! { (XTimestamp, "X-Timestamp") => [ServerTimestamp] }
 /// OAuth tokenserver api uses this instead of X-Client-State.
 header! { (XKeyID, "X-KeyID") => [String] }
 
+fn token_url(base_url: &str) -> Result<Url> {
+    let mut url = Url::parse(base_url)?;
+    // kind of gross but avoids problems if base_url has a trailing slash.
+    url.path_segments_mut()
+        // We can't do anything anyway if this is the case.
+       .map_err(|_| error::unexpected("Bad tokenserver url (cannot be base)"))?
+       .extend(&["1.0", "sync", "1.5"]);
+    Ok(url)
+}
+
+// The TokenserverToken is the token as received directly from the token server
+// and deserialized from JSON.
 #[derive(Deserialize, Clone, Debug, PartialEq, Eq)]
-pub struct TokenserverToken {
-    pub id: String,
-    pub key: String,
-    pub api_endpoint: String,
-    pub uid: u64,
-    pub duration: u64,
-    // This is treated as optional by at least the desktop client,
-    // but AFAICT it's always present.
-    pub hashed_fxa_uid: String,
+struct TokenserverToken {
+    id: String,
+    key: String,
+    api_endpoint: String,
+    uid: u64,
+    duration: u64,
+    hashed_fxa_uid: String,
 }
 
-/// This is really more of a TokenAuthenticator.
-pub struct TokenserverClient {
+// The TokenserverContext holds the TokenserverToken and anything derived
+// from that token. It supplies anything derived from the token server, such
+// as auth headers, the storage endpoint, the hashed fxa uid, etc.
+pub struct TokenserverContext {
```

Comment:
What do you think of `TokenInfo` or `TokenAuthInfo` as a name? `Context` is quite the generic name...

Reply:
Yeah, those names are better. My new names are quite different than before, and there's still a "context", but I think it makes more sense in the new patch (and its private anyway)